Assert time and audience conditions on SAML responses

mgduk opened the pull request:
SAML responses can (and normally do) supply conditions for when the assertions should be considered valid. `NotBefore` and `NotOnOrAfter` restricts to a time period. `AudienceRestriction` restricts to the intended audience.

The library was previously not validating these at all.

mgduk commented:
Fixed in the upstream saml2js repo: https://github.com/Clever/saml2/pull/79

mgduk committed:
Validate NotBefore and NotOnOrAfter conditions

SAML responses can (and normally do) contain NotBefore and NotOnOrAfter conditions, which specify when the assertion should be considered valid.

See section 2.5.1.2 of http://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf

These are now checked and if the current time is outside this window the response is rejected with an error message of 'SAML Response is not yet valid' or 'SAML Response is no longer valid'.

post_assert can now accept an option `ignore_timing` which, when `true`, skips over these checks. It's not recommended for use in production, so doesn't need to be documented, but I've used it in the existing tests to avoid changing the dates in all the existing fixtures.

mgduk committed:
Validate AudienceRestriction condition

If `options.audience` is set, and the SAML response contains an `<AudienceRestriction>` condition, at least one of the `<Audience>` nodes' value must match the `options.audience`.

The `audience` option can be either a string or a regular expression, and defaults to `entity_id` if not set. It can be passed as an option to the `ServiceProvider` constructor, or in a `post_assert()` call.

brettkiefer commented on package.json:
```diff
@@ -30,6 +30,7 @@
   "dependencies": {
     "async": "^2.5.0",
     "debug": "^2.6.0",
+    "moment": "^2.18.1",
```
Including moment for those two calls seems a little bit much -- is there a good reason?

mgduk replied:
Agreed. I've removed that and force pushed.

brettkiefer commented on lib/saml2.coffee:
```diff
@@ -595,6 +597,32 @@ module.exports.ServiceProvider =
+                audiences = audience_restriction?.getElementsByTagNameNS(XMLNS.SAML, 'Audience')
+                if audiences?.length > 0
+                  validAudience = false
+                  for audience in audiences
```
It looks like they're already including underscore .. this feels a lot like an underscore _.find -- `validAudience = _.find(ye predicate)` it'd simplify the code a little bit for the reader and modifier, and I think have all of the same performance characteristics. Would that be idiomatic within this codebase? Minor thing, feel free to brush it off if there are good reasons to do it this way.

mgduk replied:
Thanks @brettkiefer — it is better with `_.find`. I've pushed a fixup for when you're ready to re-review!

brettkiefer commented on lib/saml2.coffee:
```diff
@@ -609,18 +609,15 @@ module.exports.ServiceProvider =
 
                 audience_restriction = conditions.getElementsByTagNameNS(XMLNS.SAML, 'AudienceRestriction')[0]
-                audiences = audience_restriction?.getElementsByTagNameNS(XMLNS.SAML, 'Audience')
+                audiences =  audience_restriction?.getElementsByTagNameNS(XMLNS.SAML, 'Audience')
```
Looks like ya picked up an extra space.

mgduk replied:
Now with less space.

mgduk committed:
Add `notbefore_skew` option

Specifies a number of seconds ‘grace’ to allow when asserting an authentication response's `NotBefore` condition.

In the real world, servers clock time will not be precise.

If the IdP's clock is ahead of the SP's clock, the SP may receive a SAML authentication response with a NotBefore condition in the future (according to the SP's clock), which would be rejected.

This option allows for a tolerance of clock skew to avoid erroneously rejecting authentication responses.

It defaults to 1 (second) as a balance between security (which would prefer 0) and pragmatism (real-world clocks are likely to be slightly out from each other).

This option is consistent with other SAML implementations. For example:
 - https://github.com/onelogin/ruby-saml#clock-drift
 - http://rmichaelmead.com/adfs-not-before-time-skew/

brettkiefer commented on lib/saml2.coffee:
```diff
@@ -501,12 +501,13 @@ module.exports.ServiceProvider =
       {@entity_id, @private_key, @certificate, @assert_endpoint, @alt_private_keys, @alt_certs} = options
 
       options.audience ?= @entity_id
+      options.notbefore_skew ?= 1
```
A default of 1 instead of 0 ?

brettkiefer replied:
Oh, commit comment. OK.
